File: command.java
// public Certainly! The Command design pattern is used to encapsulate a request as an object, thereby allowing for parameterization of clients with different requests, queuing of requests, and logging of the requests. Here's a complex example of the Command pattern in Java:

// Copy
import java.util.ArrayList;
import java.util.List;

// Command interface
interface Command {
    void execute();
}

// ConcreteCommand classes
class LightOnCommand implements Command {
    private Light light;

    public LightOnCommand(Light light) {
        this.light = light;
    }

    @Override
    public void execute() {
        light.turnOn();
    }
}

class LightOffCommand implements Command {
    private Light light;

    public LightOffCommand(Light light) {
        this.light = light;
    }

    @Override
    public void execute() {
        light.turnOff();
    }
}

// Receiver class
class Light {
    public void turnOn() {
        System.out.println("Light is ON");
    }

    public void turnOff() {
        System.out.println("Light is OFF");
    }
}

// Invoker class
class RemoteControl {
    private List<Command> commandHistory = new ArrayList<>();

    public void pressButton(Command command) {
        command.execute();
        commandHistory.add(command);
    }

    public void showCommandHistory() {
        System.out.println("Command History:");
        for (Command command : commandHistory) {
            System.out.println(command.getClass().getSimpleName());
        }
    }
}

// Client code
public class CommandPatternExample {
    public static void main(String[] args) {
        // Creating receiver
        Light light = new Light();

        // Creating commands
        Command lightOnCommand = new LightOnCommand(light);
        Command lightOffCommand = new LightOffCommand(light);

        // Creating invoker
        RemoteControl remoteControl = new RemoteControl();

        // Pressing buttons and executing commands
        remoteControl.pressButton(lightOnCommand);
        remoteControl.pressButton(lightOffCommand);

        // Showing command history
        remoteControl.showCommandHistory();
    }
}
// In this example:

// Command is the command interface with the execute method.
// LightOnCommand and LightOffCommand are concrete command classes implementing the Command interface for turning the light on and off, respectively.
// Light is the receiver class that performs the actual operations.
// RemoteControl is the invoker class that holds and executes commands, and also maintains a command history.
// This example models a remote control that can turn a light on and off. Commands are encapsulated as objects, and the invoker (RemoteControl) uses these commands without knowing the details of the operations. The command history is also maintained, allowing for undo/redo functionality or other use cases. {
    

File: command.go
// Certainly! Here's a complex example of the Command design pattern in Go:

// Copy
package main

import "fmt"

// Command interface
type Command interface {
	Execute()
}

// Receiver interface
type Light interface {
	TurnOn()
	TurnOff()
}

// ConcreteReceiver
type LivingRoomLight struct{}

func (l *LivingRoomLight) TurnOn() {
	fmt.Println("Living room light is ON")
}

func (l *LivingRoomLight) TurnOff() {
	fmt.Println("Living room light is OFF")
}

// ConcreteCommand
type TurnOnCommand struct {
	light Light
}

func NewTurnOnCommand(light Light) *TurnOnCommand {
	return &TurnOnCommand{light: light}
}

func (c *TurnOnCommand) Execute() {
	c.light.TurnOn()
}

// ConcreteCommand
type TurnOffCommand struct {
	light Light
}

func NewTurnOffCommand(light Light) *TurnOffCommand {
	return &TurnOffCommand{light: light}
}

func (c *TurnOffCommand) Execute() {
	c.light.TurnOff()
}

// Invoker
type RemoteControl struct {
	command Command
}

func (rc *RemoteControl) SetCommand(command Command) {
	rc.command = command
}

func (rc *RemoteControl) PressButton() {
	rc.command.Execute()
}

func main() {
	// Receiver
	livingRoomLight := &LivingRoomLight{}

	// Concrete Commands
	turnOnCommand := NewTurnOnCommand(livingRoomLight)
	turnOffCommand := NewTurnOffCommand(livingRoomLight)

	// Invoker
	remoteControl := &RemoteControl{}

	// Pressing buttons and executing commands
	remoteControl.SetCommand(turnOnCommand)
	remoteControl.PressButton()

	remoteControl.SetCommand(turnOffCommand)
	remoteControl.PressButton()
}
In this example:

Command is the command interface with the Execute method.
Light is the receiver interface with TurnOn and TurnOff methods.
LivingRoomLight is the concrete receiver.
TurnOnCommand and TurnOffCommand are concrete command classes implementing the Command interface.
RemoteControl is the invoker class.
The program demonstrates a remote control that can turn on and off a living room light. Commands are encapsulated as objects, and the invoker (RemoteControl) uses these commands without knowing the details of the operations. The use of interfaces allows for flexibility in adding new commands and receivers without modifying the invoker.